libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m

```diff
@@ -549,679 +452,290 @@ - (void)migrateUserDefaults {
     }
-    
-    return result;
-}
 
-+ (void)setActiveUserIdentity:(SFUserAccountIdentity *)activeUserIdentity {
-    NSUserDefaults *standardDefaults = [NSUserDefaults msdkUserDefaults];
-    if (activeUserIdentity == nil) {
-        [standardDefaults removeObjectForKey:kUserDefaultsLastUserIdentityKey];
-    } else {
-        NSMutableData *auiData = [NSMutableData data];
-        NSKeyedArchiver *archiver = [[NSKeyedArchiver alloc] initForWritingWithMutableData:auiData];
-        [archiver encodeObject:activeUserIdentity forKey:kUserDefaultsLastUserIdentityKey];
-        [archiver finishEncoding];
-        [standardDefaults setObject:auiData forKey:kUserDefaultsLastUserIdentityKey];
+    [self saveAccountForUser:currentAccount error:nil];
+    [self userChanged:change];
+    return currentAccount;
+}
+
+- (SFUserAccount*) currentUser {
+    if (_currentUser == nil) {
+        [accountsLock lock];
+        NSData *resultData = nil;
+        NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
+        resultData = [userDefaults objectForKey:kUserDefaultsLastUserIdentityKey];
+        if (resultData != nil) {
+             SFUserAccountIdentity *result = nil;
+            @try {
+                NSKeyedUnarchiver *unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:resultData];
+                result = [unarchiver decodeObjectForKey:kUserDefaultsLastUserIdentityKey];
+                [unarchiver finishDecoding];
+                if(result)
+                   _currentUser = [self userAccountForUserIdentity:result];
+                else
+                    [self log:SFLogLevelError msg:@"Located current user Identity in NSUserDefaults but was not found in list of accounts managed by SFUserAccountManager."];
+            }
+            @catch (NSException *exception) {
+                [self log:SFLogLevelWarning msg:@"Could not parse current user identity from user defaults.  Setting to nil."];
+            }
+        }else {
+            [self log:SFLogLevelWarning msg:@"Could not parse locate user identity object from user defaults.  Setting to nil."];
+        }
+        [accountsLock unlock];
     }
-    [standardDefaults synchronize];
+    return _currentUser;
 }
 
-- (void)setActiveUserIdentity:(SFUserAccountIdentity *)activeUserIdentity {
-    [SFUserAccountManager setActiveUserIdentity:activeUserIdentity];
-}
+- (void)setCurrentUser:(SFUserAccount*)user {
 
-- (NSString *)activeCommunityId {
-     NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
-    return [userDefaults stringForKey:kUserDefaultsLastUserCommunityIdKey];
-}
+    BOOL userChanged = NO;
+    [accountsLock lock];
 
-- (void)setActiveCommunityId:(NSString *)activeCommunityId {
-    NSUserDefaults *userDefaults = [NSUserDefaults msdkUserDefaults];
-    if (activeCommunityId == nil) {
-        [userDefaults removeObjectForKey:kUserDefaultsLastUserCommunityIdKey];
+    if (nil==user) { //clear current user if  nil
```

Comment:
This should all be wrapped in the object reference equality check that's standard for customized property setters, i.e.:
```obj-c
if (user != _currentUser) {
    // ...
}
```